File: Backend/models/Location.js
const fs = require('fs').promises;
const path = require('path');

const DATA_FILE = path.join(__dirname, '../data/locations.json');

async function ensureDataFile() {
  try {
    await fs.access(DATA_FILE);
  } catch {
    await fs.mkdir(path.dirname(DATA_FILE), { recursive: true });
    await fs.writeFile(DATA_FILE, JSON.stringify([], null, 2));
  }
}

async function getLocations() {
  await ensureDataFile();
  const data = await fs.readFile(DATA_FILE, 'utf-8');
  return JSON.parse(data || '[]');
}

async function saveLocations(locations) {
  await ensureDataFile();
  await fs.writeFile(DATA_FILE, JSON.stringify(locations, null, 2));
}

class Location {
  constructor(data) {
    this.id = data.id || Date.now().toString();
    this.name = data.name;
    this.address = data.address;
    this.ward = data.ward || ''; // Phường/Xã
    this.district = data.district || ''; // Quận/Huyện
    this.province = data.province || ''; // Tỉnh/Thành phố
    this.latitude = data.latitude;
    this.longitude = data.longitude;
    this.level = data.level || 'district'; // 'ward', 'district', 'province'
    this.createdAt = data.createdAt || new Date().toISOString();
    this.updatedAt = data.updatedAt || new Date().toISOString();
  }

  static async findAll() {
    return await getLocations();
  }

  static async findById(id) {
    const locations = await getLocations();
    return locations.find(l => l.id === id) || null;
  }

  static async findByProvince(province) {
    const locations = await getLocations();
    return locations.filter(l => l.province === province);
  }

  static async findByDistrict(district) {
    const locations = await getLocations();
    return locations.filter(l => l.district === district);
  }

  static async create(data) {
    const locations = await getLocations();
    const location = new Location(data);
    locations.push(location);
    await saveLocations(locations);
    return location;
  }
}

module.exports = Location;

File: Backend/models/location.py
import json
from pathlib import Path
from datetime import datetime
from models.user import FileStorage


class Location:
    """Location model - file-based storage"""
    
    def __init__(self, data):
        self.id = data.get('id', str(int(datetime.now().timestamp() * 1000)))
        self.name = data.get('name')
        self.address = data.get('address')
        self.ward = data.get('ward', '')  # Phường/Xã
        self.district = data.get('district', '')  # Quận/Huyện
        self.province = data.get('province', '')  # Tỉnh/Thành phố
        self.latitude = data.get('latitude')
        self.longitude = data.get('longitude')
        self.level = data.get('level', 'district')  # ward, district, province
        self.createdAt = data.get('createdAt', datetime.now().isoformat())
        self.updatedAt = data.get('updatedAt', datetime.now().isoformat())
    
    def to_dict(self):
        """Convert to dictionary"""
        return {
            'id': self.id,
            'name': self.name,
            'address': self.address,
            'ward': self.ward,
            'district': self.district,
            'province': self.province,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'level': self.level,
            'createdAt': self.createdAt,
            'updatedAt': self.updatedAt
        }
    
    @staticmethod
    def find_all():
        """Find all locations"""
        return FileStorage.read_json('locations.json')
    
    @staticmethod
    def find_by_id(location_id):
        """Find location by ID"""
        locations = FileStorage.read_json('locations.json')
        for l in locations:
            if l['id'] == location_id:
                return l
        return None
    
    @staticmethod
    def find_by_province(province):
        """Find locations by province"""
        locations = FileStorage.read_json('locations.json')
        return [l for l in locations if l.get('province') == province]
    
    @staticmethod
    def find_by_district(district):
        """Find locations by district"""
        locations = FileStorage.read_json('locations.json')
        return [l for l in locations if l.get('district') == district]
    
    @staticmethod
    def create(data):
        """Create new location"""
        locations = FileStorage.read_json('locations.json')
        location = Location(data)
        locations.append(location.to_dict())
        FileStorage.write_json('locations.json', locations)
        
        return location.to_dict()
